### ISSUE TITLE
Double free with OpenSSL_add_all_algorithms() call

### ISSUE THREAD

**Issue by sburnwal** [openssl/openssl#18196]
I am using openssl version 1.0.2k-fips. In my application, in initialization code, I have this to init my AES key.

```
    AESData::AESData(const uint8_t *aInKey,
                     const uint8_t *aInIv,
                     const uint32_t aInBitLen)
    {
        keyBitLen = aInBitLen;
        byteLength = aInBitLen/8;
        memcpy(aesKey, aInKey, byteLength);
        memcpy(iVec, aInIv, byteLength);
        memcpy(iVCbc, aInIv, byteLength);
        /* Initialise the library */
        ERR_load_crypto_strings();
        OpenSSL_add_all_algorithms();        
    }
```

But appears to cause double free. 
```
#2  0x000074f556c11590 in DoLiveCoreDump (cause=0x74f556c197e2 "DoubleFree") at src/[heap-profiler.cc:732](https://heap-profiler.cc:732/)
#3  0x000074f556bf8d31 in (anonymous namespace)::do_free_with_callback (ptr=ptr@entry=0x5e0b0ad7fa58, 
    invalid_free_fn=0x74f556bf7270 <(anonymous namespace)::InvalidFree(void*)>) at src/[tcmalloc.cc:1299](https://tcmalloc.cc:1299/)
#4  0x000074f556c17467 in do_free (ptr=0x5e0b0ad7fa58) at src/[tcmalloc.cc:1376](https://tcmalloc.cc:1376/)
#5  tc_free (ptr=0x5e0b0ad7fa58) at src/[tcmalloc.cc:1761](https://tcmalloc.cc:1761/)
#6  0x000074f5522319fd in CRYPTO_free () from /usr/lib64/libcrypto.so.10
#7  0x000074f5522359d7 in OBJ_NAME_add () from /usr/lib64/libcrypto.so.10
#8  0x000074f5522f5a50 in OpenSSL_add_all_ciphers () from /usr/lib64/libcrypto.so.10
#9  0x000074f5522f596e in OPENSSL_add_all_algorithms_noconf () from /usr/lib64/libcrypto.so.10
```

I want to know the reason what could be causing this and how to get out of this issue?

Thanks

**Comment by paulidale** [comment]
OpenSSL 1.0.2 is no longer supported.

**Comment by petrovr** [comment]
sburnwal wrote:
> I am using openssl version 1.0.2k-fips. In my application, in initialization code, I have this to init my AES key.
>
> ```
>      ...
>          /* Initialise the library */
>          ERR_load_crypto_strings();
>          OpenSSL_add_all_algorithms();
>      }
> ```
>
> But appears to cause double free.
Load of error strings is optional. Try a) without error strings b) load errors later c) ensure initialisation only once.

Roumen

**Comment by nhorman** [comment]
Marking as inactive, to be closed when 3.4 dev cycle ends, barring further input
